src/Orleans.Core/Messaging/MessageFactory.cs

```diff
@@ -21,9 +21,32 @@ public MessageFactory(SerializationManager serializationManager, ILogger<Message
         public Message CreateMessage(InvokeMethodRequest request, InvokeMethodOptions options)
         {
             // clear transaction info if transaction operation requires new transaction.
-            ITransactionInfo transactionInfo = (options & InvokeMethodOptions.TransactionRequiresNew) == 0 ? TransactionContext.GetTransactionInfo() : null;
+            ITransactionInfo transactionInfo = TransactionContext.GetTransactionInfo();
 
-            bool isTransactionRequired = (options & InvokeMethodOptions.TransactionRequiresNew) != 0 || (options & InvokeMethodOptions.TransactionRequired) != 0;
+            // enforce mandatory transaction calls
+            if((options & InvokeMethodOptions.TransactionMandatory) != 0 && transactionInfo == null)
+            {
+                throw new NotSupportedException("Call cannot be made outside of a transaction.");
+            }
+
+            // enforce never transaction calls
+            if ((options & InvokeMethodOptions.TransactionNever) != 0 && transactionInfo != null)
+            {
+                throw new NotSupportedException("Call cannot be made within a transaction.");
+            }
+
+            // clear transaction context if creating a transaction or transaction is suppressed
+            if ((options & InvokeMethodOptions.TransactionCreate) != 0 ||
+                ((options & InvokeMethodOptions.TransactionCreateOrJoin) == 0 && 
+                 (options & InvokeMethodOptions.TransactionMandatory) == 0 &&
+                 (options & InvokeMethodOptions.TransactionSupported) == 0))
+            {
+                transactionInfo = null;
+            }
+
+            bool isTransactionRequired = (options & InvokeMethodOptions.TransactionCreate) != 0 ||
+                                         (options & InvokeMethodOptions.TransactionCreateOrJoin) != 0 ||
+                                         (options & InvokeMethodOptions.TransactionMandatory) != 0;
```

Comment:
We could probably arrange these so that it's just one bit that needs to be checked.

Reply:
Yeah, don't like the way this works.. Will look into.